### PR TITLE
Fix app bundle support

### PR DESCRIPTION
**Pull request by thalescm**
[This project](https://github.com/thalescm/buck-android-app-bundle/) shows that the current buck implementation of android app bundles is buggy, and the presented PR fixes it while adding more security and scalability to the feature, as it removes a lot of logic from buck that is already maintained by google (bundletool).

* Added bundletool to the project, bundle tool is now used to create the final bundle.
   * Bundletool correctly generates android app bundle configuration files (current impl don't)
   * Bundletool correctly adds bundletool version used (on BundleConfig.pb) so playstore knows which one to use when creating apks
   * Bundletool will validate the modules before building the final bundle (this dosn't exist on current implementation, so on success build of `android_bundle` rule can be a false positive)
* Moving module files to the right places; as documentation explains: https://developer.android.com/guide/app-bundle/
* Removed Assets,Native and BundleConfig generation steps, as this is now covered by bundletool itself.

### CODE REVIEW

**Comment by kageiit** [comment]
cc @joelmccall @styurin

**Comment by kageiit** [comment]
@styurin any comments on this?

**Comment by joelmccall** [comment]
Sorry for the lack of response, I'll take a look at this and get back to
you tomorrow

**Comment by joelmccall** [comment]
@thalescm , overall the changes look good, would you mind rebasing it to more recent commit?

**Comment by thalescm** [comment]
@joelmccall rebased :)

**Comment by joelmccall** [comment]
Builds are hitting this error with this change:
```
compile:
    [javac] Compiling 2679 source files to /home/travis/build/facebook/buck/ant-out/classes
    [javac] [compiled 590084 lines in 37867 ms: 15583.0 lines/s]
    [javac] [5021 .class files generated]
    [javac] ----------
    [javac] 1. ERROR in /home/travis/build/facebook/buck/src/com/facebook/buck/android/AabBuilderStep.java (at line 154)
    [javac] 	ApkBuilder moduleBuilder = new ApkBuilder(
    [javac]           moduleGenPath.toFile(),
    [javac]           fakeResApk,
    [javac]           null,
    [javac]           null,
    [javac]           verboseStream
    [javac]       );
    [javac] 	                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    [javac] The constructor ApkBuilder(File, File, null, null, PrintStream) is undefined
    [javac] ----------
    [javac] 1 problem (1 error)
```

**Comment by thalescm** [comment]
@joelmccall fixed the erros on Travis. Could you take another look? Tks!

File: third-party/java/aapt2-proto/BUCK [hunk 1]
```diff
@@ -0,0 +1,14 @@
+prebuilt_jar(
+    name = "prebuilt",
+    binary_jar = "aapt2-proto-0.3.1.jar",
```
Comment (styurin) [inline third-party/java/aapt2-proto/BUCK]: Please, add a readme file indicating where that file was taken from as well as the license file

File: src/com/facebook/buck/android/AabBuilderStep.java [hunk 2]
```diff
@@ -98,128 +96,132 @@ public StepExecutionResult execute(ExecutionContext context) throws IOException
...
+    ImmutableList.Builder<Path> modulesPathsBuilder = new Builder<>();
+    File fakeResApk = filesystem.createTempFile("fake", ".txt").toFile();
+    Set<String> addedFiles = new HashSet<>();
+    Set<Path> addedSourceFiles = new HashSet<>();
+
+    for (ModuleInfo moduleInfo : modulesInfo) {
+      Path moduleGenPath = getPathForModule(moduleInfo);
+      StepExecutionResult moduleBuildResult =
+          addModule(
+              context, moduleGenPath, fakeResApk, output, moduleInfo, addedFiles, addedSourceFiles);
```
Comment (styurin) [inline src/com/facebook/buck/android/AabBuilderStep.java]: `output` can be a `null` here, `addModule` need to have `@Nullable` on the corresponding argument

File: third-party/java/bundletool/BUCK [hunk 3]
```diff
@@ -1,34 +1,17 @@
...
-    deps = DEPS,
-)
-
 prebuilt_jar(
-    name = "prebuilt",
-    binary_jar = "bundle.jar",
-    licenses = ["LICENSE"],
-    source_jar = "bundle.src.jar",
+name = "prebuilt",
+    binary_jar = "bundletool-0.7.1.jar",
```
Comment (styurin) [inline third-party/java/bundletool/BUCK]: Please, keep the readme file with information where this file was taken from as well as the license file.

**Comment by thalescm** [comment]
@styurin done; Can you check if this is the expected format for the README files?

File: src/com/facebook/buck/android/AabBuilderStep.java [hunk 4]
```diff
@@ -55,40 +59,35 @@
...
       Pattern.compile("^.+\\.bc$", Pattern.CASE_INSENSITIVE);
   private final ImmutableSet<ModuleInfo> modulesInfo;
-  private final AppBuilderBase appBuilderBase;
   private final ImmutableSet<String> moduleNames;
 
   /**
    * @param modulesInfo A set of ModuleInfo containing information about modules to be built within
    *     this bundle
-   * @param pathToOutputApkFile Path to output our APK to.
+   * @param pathToOutputAabFile Path to output our APK to.
```
Comment (styurin) [inline src/com/facebook/buck/android/AabBuilderStep.java]: s/APK/AAB/

File: third-party/java/bundletool/LICENSE [hunk 5]
```diff
@@ -199,4 +199,4 @@
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
-   limitations under the License.
+   limitations under the License.
```
Comment (styurin) [inline third-party/java/bundletool/LICENSE]: Can you revert this?

**Comment by thalescm** [comment]
@styurin implemented the requested changes :)

**Comment by styurin** [comment]
Could you rebase on the latest stable?

**Comment by kageiit** [comment]
> Could you rebase on the latest stable?

Hey @styurin is there any reason why this cannot be done internally? It seems to waste precious cycles for external contributors to do this when there are no conflicts

**Comment by kageiit** [comment]
> I'm asking because there is a merge conflict caused by internal changes. I tried to change those internal pieces, but it's still failing and I hope rebasing on master would solve that. Otherwise someone would need to apply this PR manually to the internal repository which means wasting precious cycles of internal developers.

Got it. It was not clear where the conflict was. Thanks for clarifying!

**Comment by thalescm** [comment]
@styurin Rebased

**Comment by styurin** [comment]
Could you update IntelliJ project to remove the old library and add a new ones?

**Comment by thalescm** [comment]
@styurin can you check if [9d7aeed](https://github.com/facebook/buck/pull/2140/commits/9d7aeed664301bc7ec4e3a61e803e43c118c5d18) should suffice?

**Pull request closed without merging**